Update and extend documentation (#712)
pymeasure/pymeasure: docs/dev/adding_instruments.rst

> Make sure to include the PyMeasure license to each file, and add yourself as an author to the :code:`AUTHORS.txt` file.

Reviewer: ```suggestion
Make sure to include the PyMeasure license to each file, and add yourself as an author to the :code:`AUTHORS.txt` and :code:`docs/about/authors.rst` files.
```
By chance, I saw this one as well.

Author: I think I'll see if I can eliminate this nowadays needless duplication in this PR, instead.

Reviewer: That would be even better. Thanks.

Author: This is now much easier than when I tried this last time. Not super beautiful because it's a literal block instead of a bullet list, but good enough imo.

There is a certain order of elements in an instrument class that is useful to adhere to:

* First, the initializer (the :code:`__init__()` method), this makes it faster to find when browsing the source code.
* Then class attributes/variables, if you need them.
* Then properties (pymeasure-specific or generic Python variants). This will be the bulk of the implementation.

> * Finally, any methods.

Reviewer: Let us mention somewhere (I think it is fitting here or about properties), that it is perfectly fine to define methods. For example:
```suggestion
* Finally, any methods. 'Actions' like :code:`recall()` or :code:`trigger_scan()` may find their place here.
```

Author: I have some drafts for dealing with this already locally.

Author: Done.

> * Finally, any methods.

Reviewer: It is one option, to separate the properties and methods according to their type.

I, on the other hand, prefer to sort thematically, e.g. I'd add a `start_scan()` method to the `scan_speed`, `scan_range`, and `scan_center` properties.

I guess we can leave it, how you wrote it, because these methods (like `start_scan()`) are rare and can be an exception to this rule and stay with their buddy properties.

Author: I see your point. I can emphasize that it's a guideline, not a strict policy. If people start thinking about the order then it's already accomplished most of what I want. :-)

Reviewer: You wrote, that it is a useful order. Therefore it is marked as a suggestion. Keep it as is.

> 

Reviewer: Maybe this is also the place for hints about good naming patterns.
```suggestion

It's good to name properties self explaining, short, and descriptive. `voltage_setpoint`, for example, would be good for the output voltage of a voltage source, which can measure a voltage as well. Verbs are a bad choice, because a property is no action (even though it may require one to acquire the data), so the measured voltage property should be called `measured_voltage` or just `voltage`, but not `measure_voltage`.
Similarly, if you want the device to perform some action (starting/stopping something etc.), these methods (mostly without parameters and return values) should use verbs for their name, e.g. `start_scan()`.

```

Author: yes, as I said, I have some stuff locally, but it was too late yesterday. :-P
I'll set this PR to WIP or something :D

Reviewer: They're just some suggestions, that you do not have to write everything yourself.

Reviewer: Oh, great expansion. Thanks

* If there is already an instrument with a similar command set, check if you can inherit from that one and just tweak a couple of things. This massively reduces code duplication and maintenance effort. The section :ref:`instruments_with_similar_features` shows how to achieve that.
* The bulk of your instrument's interface will probably be made up of properties for quantities to set and/or read out. Our custom properties (see :ref:`writing_properties` ff. below) offer some convenience features and are therefore preferable, but plain Python properties are also fine.
* "Actions", commands or verbs should typically be methods, not properties: :code:`recall()`, :code:`trigger_scan()`, :code:`prepare_resistance_measurement()`, etc.
* This separation between properties and methods also naturally helps with observing the `"command-query separation" principle <https://en.wikipedia.org/wiki/Command%E2%80%93query_separation>`__.
* If your instrument has multiple identical channels, see XXX. TODO: write section on channel implementations

> In principle you are free to write any methods that are necessary for interacting with the instrument. When doing so, make sure to use the :code:`self.ask(command)`, :code:`self.write(command)`, and :code:`self.read()` methods to issue commands instead of calling the adapter directly. If the communication requires changes to the commands sent/received, you can override these methods in your instrument, for further information see advanced_communication_protocols_.

Reviewer: ```suggestion
In principle, you are free to write any methods that are necessary for interacting with the instrument. When doing so, make sure to use the :code:`self.ask(command)`, :code:`self.write(command)`, and :code:`self.read()` methods to issue commands instead of calling the adapter directly. If the communication requires changes to the commands sent/received, you can override these methods in your instrument, for further information see advanced_communication_protocols_.
```

In practice, we have developed a number of best practices for making instruments easy to write and maintain. The following sections detail these, which are highly encouraged to follow.

Common instrument types
***********************
There are a number of categories that many instruments fit into.
In the future, pymeasure should gain an abstraction layer based on that, see `this issue <https://github.com/pymeasure/pymeasure/issues/416>`__.
Until that is ready, here are a couple of guidelines towards a more uniform API.
Note that not all already available instruments follow these, but expect this to be harmonized in the future.

Frequent properties
-------------------
If your instrument has an **output** that can be switched on and off, use a :ref:`boolean property <boolean-properties>` called :code:`output_enabled`.

Power supplies
--------------
PSUs typically can measure the *actual* current and voltage, as well as have settings for the voltage level and the current limit.

> To keep naming clear and avoid confusion, implement the properties :code:`current`, :code:`voltage`, :code:`voltage_setpoint` and :code:`current_limit`, respectively.

Author: Here, I have started to document things I repeat again and again during reviews. I hope with this we can start shaping our instrument APIs toward homogeny before we get around to implement #416. Feel free to extend this in the future.

Reviewer: Good idea

Author: ping @msmttchr @CasperSchippers : only so that you are aware of this existing, now.

PSUs typically can measure the *actual* current and voltage, as well as have settings for the voltage level and the current limit.
To keep naming clear and avoid confusion, implement the properties :code:`current`, :code:`voltage`, :code:`voltage_setpoint` and :code:`current_limit`, respectively.

Managing status codes or other indicator values
***********************************************
Often, an instrument features one or more collections of specific values that signal some status, an instrument mode or a number of possible configuration values.
Typically, these are collected in mappings of some sort, as you want to provide a clear and understandable value to the user, while abstracting away the raw data, think :code:`ACQUISITION_MODE` instead of :code:`0x04`.

> The mappings normally are kept at module level (i.e. not defined within the instrument class), so that they are available when using the property factories.

Reviewer: Can't you access them at class level (e.g. AwG401.SomeEnum instead of self.SomeEnum)?
Because for the setter a class enum is better.

Or (for a `control`), the class may contain the module level enum.

Anyway, we can adjust the documentation according to upcoming experience.

Author: In my experience this does not work reliably (also for lambdas, etc), but we can sure test in a more systematic way. I propose to do that another time. Changing the docs to an improved approach should always be welcome. :-)

Author: See #736

.. doctest::

    >>> from enum import IntEnum
    >>> class InstrMode(IntEnum):
    ...     WAITING = 0x00
    ...     HEATING = 0x01
    ...     COOLING = 0x05
    ...
    >>> received_from_device = 0x01
    >>> current_mode = InstrMode(received_from_device)
    >>> if current_mode == InstrMode.WAITING:
    ...     print('Idle')
    ... else:
    ...     print(current_mode)
    ...     print(f'Mode value: {current_mode}')
    ...
    InstrMode.HEATING
    Mode value: 1

:code:`IntFlag` has the added benefit that it supports bitwise operators and combinations, and as such is a good fit for status bitmasks or error codes that can represent multiple values:

.. doctest::

    >>> from enum import IntFlag
    >>> class ErrorCode(IntFlag):
    ...     TEMP_OUT_OF_RANGE = 8
    ...     TEMPSENSOR_FAILURE = 4
    ...     COOLER_FAILURE = 2
    ...     HEATER_FAILURE = 1
    ...     OK = 0
    ...
    >>> received_from_device = 7
    >>> print(ErrorCode(received_from_device))
    ErrorCode.TEMPSENSOR_FAILURE|COOLER_FAILURE|HEATER_FAILURE

:code:`IntFlags` are used by many instruments for the purpose just demonstrated.

> 

Reviewer: Let's include a suggestion how to implement it:
```suggestion

The status property could look like this:
.. testcode::

    status = Instrument.measurement(
        "STB?", 
        """Measure the status of the device as enum.""",
        get_process=lambda v: ErrorCode(v), 
   )
```

Author: LGTM!

Reviewer: I just found an even better implementation:

`get_process=ErrorCode`

Sorry for the change, but we do not need a lambda for calling a method 😁

For many instruments, the simple way presented first is enough, but in case you have a more complex arrangement to implement, see whether advanced_communication_protocols_ fits your bill. If, for some exotic reason, you need a special connection type, which you cannot model with PyVISA, you can write your own Adapter.

.. _writing_properties:

Writing properties
==================

In PyMeasure, `Python properties`_ are the preferred method for dealing with variables that are read or set.

The property factories
**********************

> PyMeasure comes with three central convenience factory functions for making properties for classes: :func:`Instrument.control <pymeasure.instruments.Instrument.control>`, :func:`Instrument.measurement <pymeasure.instruments.Instrument.measurement>` and :func:`Instrument.setting <pymeasure.instruments.Instrument.setting>`.

Reviewer: In English, you have a comma before "and" in an enumeration.
```suggestion
PyMeasure comes with three central convenience factory functions for making properties for classes: :func:`Instrument.control <pymeasure.instruments.Instrument.control>`, :func:`Instrument.measurement <pymeasure.instruments.Instrument.measurement>`, and :func:`Instrument.setting <pymeasure.instruments.Instrument.setting>`.
```

Author: Well, in [Oxford](https://en.wikipedia.org/wiki/Serial_comma), at least. :wink: 
(for the record, I also use the Oxford comma, just missed that one)

The :func:`Instrument.measurement <pymeasure.instruments.Instrument.measurement>` function returns a property that can only read values from an instrument.

> For example, if our "Extreme 5000" has the :code:`*IDN?` command we can write the following property to be added after the :code:`def __init__` line in our above example class, or added to the class after the fact as in the code here:

Reviewer: ```suggestion
For example, if our "Extreme 5000" has the :code:`*IDN?` command, we can write the following property to be added after the :code:`def __init__` line in our above example class, or added to the class after the fact as in the code here:
```

     )

.. testcode::
    :hide:
    
    # We have to fake this silently because the FakeInstrument cannot do
    # a measurement property, it only mirrors values that you sent first.
    Extreme5000.cell_temp = 127.2
    
You will notice that a documentation string is required, see :ref:`docstrings` for details.

When we use this property we will get the temperature of the reaction cell.

.. doctest::

    >>> extreme = Extreme5000("GPIB::1")
    >>> extreme.cell_temp  # Sends ":TEMP?" to the device
    127.2

The :func:`Instrument.control <pymeasure.instruments.Instrument.control>` function extends this behavior by creating a property that you can read and set. For example, if our "Extreme 5000" has the :code:`:VOLT?` and :code:`:VOLT <float>` commands that are in Volts, we can write the following property.

.. testcode::

    Extreme5000.voltage = Instrument.control(
        ":VOLT?", ":VOLT %g",
        """Control the voltage in Volts (float)."""
    )

You will notice that we use the `Python string format`_ :code:`%g` to format passed-through values as floating point.

.. _Python string format: https://docs.python.org/3/library/string.html#format-specification-mini-language

> We can use this property to set the voltage to 100 mV, which will send the appropriate command and then requests the current voltage.

Reviewer: Sorry for correcting again this sentence to make it more clear, that you describe two different actions of this `control`:
```suggestion
We can use this property to set the voltage to 100 mV, which will send the appropriate command, and then to request the current voltage:
```

Author: No worries, it's clearer.

> We can use this property to set the voltage to 100 mV, which will sends the appropriate command and then requests the current voltage.

Reviewer: ```suggestion
We can use this property to set the voltage to 100 mV, which will send the appropriate command and then requests the current voltage.
```

.. doctest::

    >>> extreme = Extreme5000("GPIB::1")

>     >>> extreme.voltage = 0.1        # Sends ":VOLT 0.1"

Reviewer: Mirroring the second command.
```suggestion
    >>> extreme.voltage = 0.1        # Sends ":VOLT 0.1" to set the voltage to 100 mV
```

    >>> extreme.voltage              # Sends ":VOLT?" to query for the current value
    0.1

Finally, the :func:`Instrument.setting <pymeasure.instruments.Instrument.setting>` function can only set, but not read values.

> Using the :func:`Instrument.control <pymeasure.instruments.Instrument.control>`, :func:`Instrument.measurement <pymeasure.instruments.Instrument.measurement>` and :func:`Instrument.control <pymeasure.instruments.Instrument.control>` functions, you can create a number of properties for basic measurements and controls.

Reviewer: ```suggestion
Using the :func:`Instrument.control <pymeasure.instruments.Instrument.control>`, :func:`Instrument.measurement <pymeasure.instruments.Instrument.measurement>`, and :func:`Instrument.control <pymeasure.instruments.Instrument.control>` functions, you can create a number of properties for basic measurements and controls.
```

> The next sections detail additional features of :func:`Instrument.control <pymeasure.instruments.Instrument.control>` that allow you to write properties that cover specific ranges, or have to map between a real value to one used in the command. Furthermore it is shown how to perform more complex processing of return values from your device.

Reviewer: including the derivatives as well or does the user understand, that `measurement` and `setting` work the same?
```suggestion
The next sections detail additional features of :func:`Instrument.control <pymeasure.instruments.Instrument.control>` and derivatives that allow you to write properties that cover specific ranges, or have to map between a real value to one used in the command. Furthermore it is shown how to perform more complex processing of return values from your device.
```

Author: Noted, I instead used the term "property factories" from the previous heading.

As you have seen, the :func:`Instrument.control <pymeasure.instruments.Instrument.control>` function can be significantly extended by using validators and maps.

Boolean properties
******************

The idea of using maps can be leveraged to implement properties where the user-facing values are booleans, so you can interact in a pythonic way using :code:`True` and :code:`False`:

.. testcode::

    Extreme5000.output_enabled = Instrument.control(
        "OUTP?", "OUTP %d",
        """Control the instrument output is enabled (boolean).""",
        map_values=True,
        values={True: 1, False: 0},  # the dict values could also be "on" and "off", etc.
    )

.. doctest::

    >>> extreme = Extreme5000("GPIB::1")
    >>> extreme.output_enabled = True
    >>> extreme.read()
    '1'
    >>> extreme.output_enabled = False
    >>> extreme.output_enabled
    False
    >>> # Invalid input raises an exception
    >>> extreme.output_enabled = 34
    Traceback (most recent call last):
    ...

>     KeyError: 34

Author: So, I realised that we don't need the `strict_discrete_set` validator for the boolean property -- if an invalid input is supplied, you get a KeyError from the dict lookup as shown in the example.

Do you think we should leave it in the docs like this, or with the `strict_discrete_set` added, as the error message might be clearer/more useful?

Reviewer: I personally prefer cleaner code (with less additions).

However, a validator ensures the right values. 

We could add, that a mapping dictionary works as a validator as well (maybe even transform the exception in `values` from KeyError to Value Error("Invalid value supplied for mapping") or something.

This should enter the docstring and documentation, I'd say.

Author: > We could add, that a mapping dictionary works as a validator as well (maybe even transform the exception in values from KeyError to Value Error("Invalid value supplied for mapping") or something.

That's probably more confusing to users. Let's add the `strict_discrete_set` back, the error message is clearer, and semantically it's also clearer that the inputs are being validated (even if they would also be "validated" in a way by the subsequent dict lookup).

Author: I put the validator in now. Agreed to do it that way?
(Note: I know that the error message talks about a set but displays a dict, but that's a battle for another day)

Reviewer: In the end, it is an example how you can do it. With the validator in the example, the readers will see how to validate correctly.

So, I'm ok with the validator.

As you have seen, the :func:`Instrument.control <pymeasure.instruments.Instrument.control>` function can be significantly extended by using validators and maps.

.. _boolean-properties:

Boolean properties
******************

The idea of using maps can be leveraged to implement properties where the user-facing values are booleans, so you can interact in a pythonic way using :code:`True` and :code:`False`:

.. testcode::

    Extreme5000.output_enabled = Instrument.control(
        "OUTP?", "OUTP %d",
        """Control the instrument output is enabled (boolean).""",
        map_values=True,

>         values={True: 1, False: 0},  # the dict values could also be "on" and "off", etc.

Reviewer: ```suggestion
        values={True: 1, False: 0},  # the dict values could also be "on" and "off", etc. depending on the device
```

    )

.. doctest::

    >>> extreme = Extreme5000("GPIB::1")
    >>> extreme.output_enabled = True

>     >>> extreme.read()
>     '1'
>     >>> extreme.output_enabled = False
>     >>> extreme.output_enabled
>     False

Reviewer: How can you read without prior writing?
I guess you ment to show the raw communication.
```suggestion
    >>> extreme.write("OUTP?")
    >>> extreme.read()  # read the response of the device prior to mapping.
    '1'
    >>> extreme.output_enabled = False
    >>> extreme.output_enabled  # read the mapped value
    False
```

Author: Indeed. I think this is a bit of an artifact of how our faked instrument works like, and occurs a couple of times more.

Author: This will need some deeper changes if we want it to be more representative -- the first line `extreme.output_enabled = True` basically puts a (mapped) `1` into a primitive buffer in `FakeAdapter`, which the `read` consumes without mapping.

If we can't easily adapt `FakeAdapter` to give more meaningful (doctestable) interaction, maybe the easiest thing would be to define a pyvisa-sim device for the handful of examples that use that misleading raw `read()` approach.

Author: I'll bump this to another issue, we have been doing this for a long time that way and nobody noticed, and it will just drag this PR out much longer :D

Author: I'll resolve this in favour of #735.

Reviewer: Ok, resolve it. Somebody has to know the code to find the error, but those don't need the example for the basic communication.

    Extreme5000.capacity = Instrument.measurement(
        ":CAP?",
        """Measure the capacity in nF (float).""",
        preprocess_reply=lambda v: v.replace('nF', '')
        # notice how we don't need to cast to float anymore
    )

Tweaking command strings
************************
If you need to tweak

* the :code:`set_command` string immediately before the value to set is inserted via string formatting (:code:`%g` etc.), or
* the :code:`get_command` string before sending it to the device,

use the :code:`command_process` parameter of :meth:`~pymeasure.instruments.Instrument.control`.

> Note that there is only parameter for both setting and getting, so the utility of this is probably limited.
> Note also that for adding e.g. channel identifiers, there are other, more preferable methods.

Reviewer: ```suggestion
Note, that there is only one parameter for both setting and getting, so the utility of this is probably limited.
Note also, that for adding e.g. channel identifiers, there are other, more preferable methods.
```

Author: No comma here, either, "note that the sky is blue" is one phrase. I added the "one", though. :-P

Reviewer: You're right.

Now our voltage property has a dynamic validity range, either [-1, 1] or [0, 1].

> A side effect of this is that the property's docstring should be less specific, to avoid it containing dynamically changed information (like the admissible value range).

Reviewer: ```suggestion
A side effect of this is, that the property's docstring should be less specific, to avoid it containing dynamically changed information (like the admissible value range).
```

Author: I don't agree with all comma suggestions, this one for example I'm pretty sure is non appropriate.

In this example, the property name was :code:`voltage` and the parameter to adjust was :code:`values`, so we used :code:`self.voltage_values` to set our desired values.

.. _instruments_with_similar_features:

Instruments with similar features
=================================

When instruments have a similar set of features, it makes sense to use inheritance to obtain most of the functionality from a parent instrument class, instead of copy-pasting code.

.. note::
    Don't forget to update the instrument's :code:`name` attribute accordingly, by either supplying an appropriate argument (if available) during the :code:`super().__init__()` call, or by setting it anew below that call.

> Sometimes one only needs to add additional properties and methods.
> Often, some of the already present properties/methods need to be completely replaced by defining them again in the derived class.
> Often, however, only some details need to be changed.
> This can be dealt with efficiently using dynamic properties.

Reviewer: I'm not happy with the duplication of "often"
```suggestion
In some cases, one only needs to add additional properties and methods.
In other cases, some of the already present properties/methods need to be completely replaced by defining them again in the derived class.
Often, however, only some details need to be changed.
This can be dealt with efficiently using dynamic properties.
```
